src/lib/yang/xpath.lua

```diff
@@ -0,0 +1,281 @@
+-- named keys such as [addr=1.2.3.4] and also basic positional querying
+-- for arrays e.g [position()=1] for the first element.
+--
+-- The structure of the path is dependent on the type the node is. The
+-- conversions are as follows:
+--
+-- Scalar fields:
+--    A lua string of the member name
+-- Struct fields:
+--    A lua string of the member name
+-- Array fields:
+--    This is a table which has a "name" property specifying member
+--    name and a "key" field which is a 1 based integer to specify the
+--    position in the array.
+-- Table fields:
+--    This is a table which has a "name" property specifying member
+--    name and has a "keys" (not key) property which is either:
+--       - A string representing the key if the table is string keyed.
+--       - A lua table with corrisponding leaf names as the key and the
+--         value as the value.
+module(..., package.seeall)
+
+local ffi = require("ffi")
+local lib = require("core.lib")
+local schemalib = require("lib.yang.schema")
+local datalib = require("lib.yang.data")
+local valuelib = require("lib.yang.value")
+local util = require("lib.yang.util")
+local normalize_id = datalib.normalize_id
+
+local function extract_parts(fragment)
+   local rtn = {query={}}
+   rtn.name = string.match(fragment, "([^%[]+)")
+   for k,v in string.gmatch(fragment, "%[([^=]+)=([^%]]+)%]") do
+      rtn.query[k] = v
+   end
+   return rtn
+end
+
+local handlers = {}
+function handlers.scalar(fragment, tree)
+   return fragment.name, tree
+end
+function handlers.struct(fragment, tree)
+   return fragment.name, tree.members
+end
+function handlers.table(fragment, tree)
+   if #tree.keys == 1 then
+      -- Check if the key is a string based key
+      if tree.string_key then
+         return {name=fragment.name, key=tree.string_key}, tree
+      end
+   else
+      return {name=fragment.name, keys=fragment.query}, tree
+   end
+end
+function handlers.array(fragment, tree)
+   local position = fragment.query["position()"]
+   return {name=fragment.name, key=tonumber(position)}
+end
+function handle(node_type, fragment, tree)
+   return assert(handlers[node_type], node_type)(fragment, tree)
+end
+
+-- Gets the next item in the path returning the element and the remaining
+-- path fragment. For example "router.routes.route" will return "router"
+-- and "routes.route". If the end is reached it'll return nil.
+function next_element(path)
+   return string.match(path, "([^/]+)/?(.*)")
+end
+
+-- Converts an XPath path to a lua array consisting of path componants.
+-- A path component can then be resolved on a yang data tree:
+
+local function convert_path(path, grammar)
+   -- Extract head, check type and dispatch to handler.
+   local head, tail = next_element(path)
+   local parts = extract_parts(head)
+   local err = "Invalid path: "..parts.name
+   local node
+   if grammar.type == "table" then
+      if grammar.keys[head] == nil then
+         node = assert(grammar.values[parts.name], err)
```

Comment:
This is to support looking inside of lists, to use an example from the tests "/routes/route[addr=1.2.3.4]/port" it would be the "port". We don't support list selection via values, only by keys.